File: modulo08-arquitetura-de-sistemas-com-ia/modulo-04-padroes-ai-especificos/trialforge-gateway-prototype-java/src/main/java/com/trialforge/gateway/Bm25.java
package com.trialforge.gateway;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * BM25 classico (Hybrid Search, Modulo 4.1) — a metade lexical da fusao com
 * o embedding denso.
 */
public final class Bm25 {

    private Bm25() {
    }

    /** Estatisticas do corpus, pre-computadas uma vez na indexacao. */
    public record EstatisticasBM25(List<List<String>> tokensPorDoc, int n, double avgdl, Map<String, Integer> df) {
    }

    public static EstatisticasBM25 construirEstatisticasBM25(List<String> documentos) {
        List<List<String>> tokensPorDoc = new ArrayList<>();
        int somaComprimentos = 0;
        for (String doc : documentos) {
            List<String> tokens = Tokenizer.tokenizar(doc);
            tokensPorDoc.add(tokens);
            somaComprimentos += tokens.size();
        }
        int n = documentos.size();
        double avgdl = (double) somaComprimentos / n;

        Map<String, Integer> df = new HashMap<>();
        for (List<String> tokens : tokensPorDoc) {
            Set<String> vistos = new HashSet<>(tokens);
            for (String termo : vistos) {
                df.merge(termo, 1, Integer::sum);
            }
        }

        return new EstatisticasBM25(tokensPorDoc, n, avgdl, df);
    }

    public static double scoreBM25(List<String> queryTokens, List<String> docTokens, EstatisticasBM25 estatisticas) {
        return scoreBM25(queryTokens, docTokens, estatisticas, 1.5, 0.75);
    }

    public static double scoreBM25(List<String> queryTokens, List<String> docTokens, EstatisticasBM25 estatisticas,
            double k1, double b) {
        double score = 0;
        for (String termo : queryTokens) {
            long freqNoDoc = docTokens.stream().filter(t -> t.equals(termo)).count();
            if (freqNoDoc == 0) {
                continue;
            }
            int docFreq = estatisticas.df().getOrDefault(termo, 0);
            // +1 evita idf negativo com poucos docs
            double idf = Math.log((estatisticas.n() - docFreq + 0.5) / (docFreq + 0.5) + 1);
            double numerador = freqNoDoc * (k1 + 1);
            double denominador = freqNoDoc + k1 * (1 - b + (b * docTokens.size()) / estatisticas.avgdl());
            score += idf * (numerador / denominador);
        }
        return score;
    }
}

File: modulo08-arquitetura-de-sistemas-com-ia/modulo-04-padroes-ai-especificos/trialforge-gateway-prototype-go/gateway/bm25.go
package gateway

import "math"

// EstatisticasBM25 guarda o que é pré-computado uma vez por corpus: tokens de
// cada documento, N, comprimento médio (avgdl) e document frequency por termo.
type EstatisticasBM25 struct {
	TokensPorDoc [][]string
	N            int
	Avgdl        float64
	DF           map[string]int
}

// ConstruirEstatisticasBM25 pré-computa as estatísticas do corpus — feito uma
// vez na indexação, nunca a cada busca.
func ConstruirEstatisticasBM25(documentos []string) EstatisticasBM25 {
	tokensPorDoc := make([][]string, len(documentos))
	somaComprimentos := 0
	for i, doc := range documentos {
		tokensPorDoc[i] = Tokenizar(doc)
		somaComprimentos += len(tokensPorDoc[i])
	}
	n := len(documentos)
	avgdl := float64(somaComprimentos) / float64(n)

	df := make(map[string]int)
	for _, tokens := range tokensPorDoc {
		vistos := make(map[string]bool)
		for _, termo := range tokens {
			if vistos[termo] {
				continue
			}
			vistos[termo] = true
			df[termo]++
		}
	}

	return EstatisticasBM25{TokensPorDoc: tokensPorDoc, N: n, Avgdl: avgdl, DF: df}
}

// ScoreBM25 calcula o score BM25 clássico (k1=1.5, b=0.75 por padrão) de um
// documento (já tokenizado) contra os tokens da query.
func ScoreBM25(queryTokens, docTokens []string, estatisticas EstatisticasBM25, k1, b float64) float64 {
	score := 0.0
	for _, termo := range queryTokens {
		freqNoDoc := 0
		for _, t := range docTokens {
			if t == termo {
				freqNoDoc++
			}
		}
		if freqNoDoc == 0 {
			continue
		}
		docFreq := estatisticas.DF[termo]
		// +1 evita idf negativo com poucos docs
		idf := math.Log((float64(estatisticas.N)-float64(docFreq)+0.5)/(float64(docFreq)+0.5) + 1)
		numerador := float64(freqNoDoc) * (k1 + 1)
		denominador := float64(freqNoDoc) + k1*(1-b+(b*float64(len(docTokens)))/estatisticas.Avgdl)
		score += idf * (numerador / denominador)
	}
	return score
}

// ScoreBM25Padrao chama ScoreBM25 com os defaults k1=1.5, b=0.75 usados nos
// dois originais.
func ScoreBM25Padrao(queryTokens, docTokens []string, estatisticas EstatisticasBM25) float64 {
	return ScoreBM25(queryTokens, docTokens, estatisticas, 1.5, 0.75)
}
